### PR TITLE
Coco

### LINKED ISSUE #858: Efficientdet-d0 pre-trained model not achieving advertised accuracy?
Going through the steps as laid out in the README, in the [Eval on COCO 2017](https://github.com/google/automl/tree/master/efficientdet#7-eval-on-coco-2017-val-or-test-dev) section, I achieve the following accuracy. 

```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.300
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.441
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.330
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.081
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.340
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.503
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.290
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.361
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.362
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.096
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.407
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.606
```

This is not in line with the expected accuracy, as advertised in the [d0 checkpoint validation output file](https://storage.googleapis.com/cloud-tpu-checkpoints/efficientdet/coco2/val/d0_coco_val.txt):

```
{'AP': 0.34292528, 'AP50': 0.52428037, 'AP75': 0.36577332, 'APl': 0.53778183, 'APm': 0.39954725, 'APs': 0.13054028, 'ARl': 0.7062811, 'ARm': 0.5674938, 'ARmax1': 0.29450703, 'ARmax10': 0.45863748, 'ARmax100': 0.48204282, 'ARs': 0.20185822, 'box_loss': 0.003652954, 'cls_loss': 0.35299847, 'loss': 0.6351326}
```

In the advertised checkpoint file, the AP is 34.3%, while my evaluation only achieves 30.0%.

### PR DESCRIPTION
**Pull request by fsx950223**
Fix #858

### CODE REVIEW

File: efficientdet/keras/train_lib_test.py [hunk 1]
```diff
@@ -32,11 +32,13 @@ def test_display_callback(self):
...
     tf.io.write_file(sample_image, fake_jpeg)
-    display_callback = train_lib.DisplayCallback(sample_image, config.model_dir)
+    display_callback = train_lib.DisplayCallback(sample_image,
+                                                 config.model_dir,
+                                                 1)
     model = train_lib.EfficientDetNetTrain(config=config)
     model.build((1, 512, 512, 3))
     display_callback.set_model(model)
-    display_callback.on_epoch_end(0, {})
+    display_callback.on_train_batch_end(1, {})
```
Comment (mingxingtan) [inline efficientdet/keras/train_lib_test.py:41]: what does it mean to change 0 to 1?

File: efficientdet/keras/train_lib_test.py [hunk 2]
```diff
@@ -32,11 +32,13 @@ def test_display_callback(self):
     fake_jpeg = tf.image.encode_jpeg(fake_image)
     sample_image = 'ram://fake_image.jpg'
     tf.io.write_file(sample_image, fake_jpeg)
-    display_callback = train_lib.DisplayCallback(sample_image, config.model_dir)
+    display_callback = train_lib.DisplayCallback(sample_image,
+                                                 config.model_dir,
+                                                 1)
```
Comment (mingxingtan) [inline efficientdet/keras/train_lib_test.py:37]: how about "update_freq=1"  to help readability?

File: efficientdet/keras/train_lib.py [hunk 3]
```diff
@@ -309,6 +309,52 @@ def get_optimizer(params):
...
+    label_map = label_util.get_label_map(config.label_map)
+    log_dir = os.path.join(config.model_dir, 'coco')
+    self.file_writer = tf.summary.create_file_writer(log_dir)
+    self.evaluator = coco_metric.EvaluationMetric(
+        filename=config.val_json_file, label_map=label_map)
+
+  @tf.function
+  def _update_map(self, images, labels):
+    from keras import postprocess
+    cls_outputs, box_outputs = self.model(images, training=False)
```
Comment (mingxingtan) [inline efficientdet/keras/train_lib.py:335]: Move "from keras import postprocess" to the top?

File: efficientdet/det_model_fn.py [hunk 4]
```diff
@@ -424,10 +425,13 @@ def model_fn(inputs):
 
     def metric_fn(**kwargs):
       """Returns a dictionary that has the evaluation metrics."""
+      global params
+      params = copy.deepcopy(params)
+      params['nms_configs']['score_thresh'] = 1e-3
```
Comment (mingxingtan) [inline efficientdet/det_model_fn.py:430]: maybe just set score_thresh = 0 as default in hparams_config.py? (then you don't need to manually set score_thresh here).

The current default value 0.4 is for certain inference cases, but in reality, people should set different score_thresh for their dataset. 

WDYT?

File: efficientdet/hparams_config.py [hunk 5]
```diff
@@ -183,6 +186,7 @@ def default_detection_configs():
   h.autoaugment_policy = None
   h.grid_mask = False
   h.sample_image = None
+  h.map_freq = 5
```
Comment (mingxingtan) [inline efficientdet/hparams_config.py:189]: Maybe add a comment like:

    h.map_freq = 5     # AP eval frequency in epochs.

File: efficientdet/keras/train_lib.py [hunk 6]
```diff
@@ -309,6 +309,52 @@ def get_optimizer(params):
...
+  def __init__(self, test_dataset, update_freq=None):
+    super().__init__()
+    self.test_dataset = iter(test_dataset)
+    self.update_freq = update_freq
+
+  def set_model(self, model: tf.keras.Model):
+    import coco_metric
+    from keras import label_util
+    self.model = model
+    config = copy.deepcopy(model.config)
```
Comment (mingxingtan) [inline efficientdet/keras/train_lib.py:322]: if default value is for eval, then you don't need to change the value here.

**Review by mingxingtan — COMMENTED** [review]
Looks great. Thanks @fsx950223

File: efficientdet/det_model_fn.py (hunk 4 above)
Comment (fsx950223) [reply efficientdet/det_model_fn.py]: I agree.

File: efficientdet/keras/train_lib_test.py [hunk 7]
```diff
@@ -32,11 +32,13 @@ def test_display_callback(self):
...
     tf.io.write_file(sample_image, fake_jpeg)
-    display_callback = train_lib.DisplayCallback(sample_image, config.model_dir)
+    display_callback = train_lib.DisplayCallback(sample_image,
+                                                 config.model_dir,
+                                                 update_freq=1)
     model = train_lib.EfficientDetNetTrain(config=config)
     model.build((1, 512, 512, 3))
     display_callback.set_model(model)
-    display_callback.on_epoch_end(0, {})
+    display_callback.on_train_batch_end(0, {})
```
Comment (mingxingtan) [inline efficientdet/keras/train_lib_test.py:41]: I believe you can just use display_callback.on_train_batch_end(batch=0) ?

logs={} should be equivalent to logs=None (default)

**Review by mingxingtan — APPROVED** [review]
